File: Count Inversions - GFG/count-inversions.py
class Solution:
    #User function Template for python3
    
    # arr[]: Input Array
    # N : Size of the Array arr[]
    #Function to count inversions in the array.
    def inversionCount(self, arr, n):
        temp = [0]*n
        return count_merge(arr,temp,0,n-1)

def merge(arr,temp,left,mid,right):
    i = left
    k = left
    j = mid
    
    inv_count = 0

    while i <= mid - 1 and j <= right:
        if arr[i] <= arr[j]:
            temp[k] = arr[i]
            i+=1
            k+=1
        else:
            temp[k] = arr[j]
            j+=1
            k+=1
            
            inv_count += mid - i
            
    while i <= mid - 1:
        temp[k] = arr[i]
        i+=1
        k+=1
    
    while j <= right:
        temp[k] = arr[j]
        j+=1
        k+=1
    
    for i in range(left,right+1):
        arr[i] = temp[i]
        
            
    return inv_count

def count_merge(arr, temp, left, right):
    inv_count = 0
    
    if left < right:
        mid = (left+right)//2
        inv_count += count_merge(arr,temp,left,mid)
        inv_count += count_merge(arr,temp,mid+1,right)
        
        inv_count += merge(arr,temp,left,mid+1,right)
    
    return inv_count

#{ 
#  Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())

if __name__=='__main__':
    t = int(input())
    for tt in range(t):
        n = int(input())
        a = list(map(int, input().strip().split()))
        obj = Solution()
        print(obj.inversionCount(a,n))
# } Driver Code Ends

File: Count Inversions - GFG/count-inversions.java
// { Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class Sorting
{
    public static void main (String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        long t = sc.nextLong();
        
        while(t-- > 0)
        {
            long n = sc.nextLong();
            long arr[] = new long[(int)n];
            
            for(long i = 0; i < n; i++)
             arr[(int)i] = sc.nextLong();
             
            System.out.println(new Solution().inversionCount(arr, n));
            
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    // arr[]: Input Array
    // N : Size of the Array arr[]
    //Function to count inversions in the array.
    static long inversionCount(long arr[], long N){
        long[] temp = new long[(int)N];
        return countMergeSort(arr,temp,0,(int)N-1);
    }
    
    static long countMergeSort(long arr[], long temp[], int left, int right){
        int mid;
        
        long invCount = 0;
        if(right > left){
            mid = (right+left)/2;
            invCount += countMergeSort(arr,temp,left,mid);
            invCount += countMergeSort(arr,temp,mid+1,right);
            
            //merge
            invCount+= merge(arr,temp,left,mid+1,right);
        }
        return invCount;
    }
    
    static long merge(long arr[], long temp[], int left, int mid, int right){
        int i=left,j=mid,k=left;
        
        long invCount = 0;
        
        while(i<= mid-1 && j <= right){
            if(arr[i] <= arr[j]){
                temp[k++] = arr[i++];
            } else{
                temp[k++] = arr[j++];
                
                //if a[i] > a[j] then a[i]...a[mid] > a[j]
                invCount += (mid - i);
            }
        }
        
        while(i <= mid -1){
            temp[k++] = arr[i++];
        }
        
        while(j <= right){
            temp[k++] = arr[j++];
        }
        
        //copying merged elements??
        for(i = left; i <= right; i++){
            arr[i] = temp[i];
        }
        
        return invCount;
    }
}